### PR TITLE
Add partial solutions to the homework

### PR DESCRIPTION
**Pull request by rajgthub**
I have done the exercises in the folder I and II. I will do test of them  later. Please have a look! Thanks for your time.

### CODE REVIEW

File: I.write-tests/05-get-second-third-smallest/get-second-third.test.js [hunk 1]
```diff
@@ -1,5 +1,16 @@
...
+const inputLenAfterPassed = input.length;
 
-// also test that the original array has not changed
\ No newline at end of file
+test('should get second and third higerst numbers', () =>{
+    expect(result).toEqual([6, 8]);
+});
+// also test that the original array has not changed
+test('should not change original input array', () => {
+    expect(inputLenAfterPassed).toBe(inputLenInital);
```
Comment (40thieves) [inline I.write-tests/05-get-second-third-smallest/get-second-third.test.js]: Unfortunately this test doesn't quite do what you want it to. For example the implementation code may mutate the array that it receives as a parameter, so just checking the length isn't enough unfortunately.

To check that an array is in the same order as it was before, you would have to do something like:

```js
expect(input).toEqual([90, 5, 11, 8, 6])
```

I know this looks a bit strange as you would think that you could just replace the array here with the input variable. However due to the way *memory* works in JS, `input` might not be the same as that array after the test has run.

File: II.pass-tests/03-sum-numbers/add-numbers.js [hunk 2]
```diff
@@ -1,4 +1 @@
-function addNumbers(numbers) {
-
-}
-module.exports = addNumbers;
\ No newline at end of file
+module.exports = numbers => numbers.reduce((a, b) => a + b);
```
Comment (40thieves) [inline II.pass-tests/03-sum-numbers/add-numbers.js]: Nice 😎 There is 1 little edge case that might be problematic here though - what happens if you pass an empty array (`[]`) to this function? How could you fix it?

File: II.pass-tests/07-second-largest/second-largest.test.js [hunk 3]
```diff
@@ -5,5 +5,5 @@ test('Second largest', function () {
 
     var output = secondLargest(numbers);
 
-    expect(output).toEqual(2);
+    expect(output).toEqual(23);
```
Comment (40thieves) [inline II.pass-tests/07-second-largest/second-largest.test.js]: Sorry! Another mistake 😞 I'll make sure it's fixed for next time

File: II.pass-tests/08-get-average/get-average.js [hunk 4]
```diff
@@ -1,3 +1,8 @@
 // the input is an array of numbers and strings
 // return the average of all the numbers
-// be sure to exclude the strings
\ No newline at end of file
+// be sure to exclude the strings
+exports.average = elements => {
```
Comment (40thieves) [inline II.pass-tests/08-get-average/get-average.js]: I think the recommended way to export a function in Node is to use `module.exports = YOUR_EXPORT`

File: II.pass-tests/08-get-average/get-average.js [hunk 5]
```diff
@@ -1,3 +1,8 @@
...
+    const numbers = elements.filter(element => typeof element === "number")
```
Comment (40thieves) [inline II.pass-tests/08-get-average/get-average.js]: Nice job checking the types 😄

File: II.pass-tests/08-get-average/get-average.js [hunk 6]
```diff
@@ -1,3 +1,8 @@
...
+    return numbers.reduce((a, b) => a + b) / numbers.length;
```
Comment (40thieves) [inline II.pass-tests/08-get-average/get-average.js]: Like the exercise above, this function might have a problem if you pass an empty array

File: II.pass-tests/08-get-average/get-average.test.js [hunk 7]
```diff
@@ -1,4 +1,4 @@
-var average = require('./get-average');
+var {average} = require('./get-average');
```
Comment (40thieves) [inline II.pass-tests/08-get-average/get-average.test.js]: If you switch to using `module.exports = YOUR_EXPORT` in the implementation, you won't need to change this.

File: II.pass-tests/ES-6-practice/es6.js [hunk 8]
```diff
@@ -1,33 +1,33 @@
-// Turn this function into an arrow function
```
Comment (40thieves) [inline II.pass-tests/ES-6-practice/es6.js]: Sorry I wasn't in class last week - did you cover ES6? If not then I think this will be upcoming homework - it's fine to do it now, but be aware that there may be less homework for you to do in the future (because you've already done it).

File: II.pass-tests/ES-6-practice/es6.js [hunk 9]
```diff
@@ -1,33 +1,33 @@
...
 
 module.exports = {
-  arrow,
-  defaultParameter,
-  combineArrays,
-  destructuring,
-  templateString
+  arrow: () => "es6 is awesome!",
+  defaultParameter: (name = 'sam') = name,
+  combineArrays: (arr1, arr2) => arr1.concat(arr2),
```
Comment (40thieves) [inline II.pass-tests/ES-6-practice/es6.js]: The exercise asked that you use a spread operator (`...`) here. How could you change this implementation to use this operator?

File: II.pass-tests/ES-6-practice/es6.js [hunk 10]
```diff
@@ -1,33 +1,33 @@
...
+  destructuring : (obj = {}) => obj.cyf,
```
Comment (40thieves) [inline II.pass-tests/ES-6-practice/es6.js]: This isn't quite right. The exercise ask you to use *destructuring*, but this is using the dot style.

File: II.pass-tests/ES-6-practice/es6.js (hunk 8 above)
Comment (40thieves) [reply II.pass-tests/ES-6-practice/es6.js]: We'll cover ES6 in more depth in future classes.

File: II.pass-tests/ES-6-practice/es6.js [hunk 11]
```diff
@@ -1,33 +1,33 @@
...
+  templateString: (a, b) => "The sum is equal to " + (a + b).toString()
```
Comment (40thieves) [inline II.pass-tests/ES-6-practice/es6.js]: This also isn't quite right. The exercise asked you to use a template string (with the ` character)

File: I.write-tests/05-get-second-third-smallest/get-second-third.js [hunk 12]
```diff
@@ -1,5 +1,5 @@
 module.exports = function (array) {
-    array.sort(function (x, y) {
+    const newArray = array.sort(function (x, y) {
```
[the remark this answers is not in the document]
Comment (rajgthub) [reply I.write-tests/05-get-second-third-smallest/get-second-third.js]: Yes, you are right. I have fixed it! 
module.exports = function (array) {
    const [, secondHighest, thirdHighest] = array
      .concat()
      .sort(function(x, y) {
        return x > y;
      });
    return [secondHighest, thirdHighest];
};

[the remark this answers is not in the document]
Comment (40thieves) [reply I.write-tests/05-get-second-third-smallest/get-second-third.js]: Huh, that is an interesting approach! I'm not sure I would have thought of that 😄 There is another way which I think is a bit more *idiomatic* (which means that most JS programmers would do it in a certain way):

```js
const newArray = array.slice()
newArray.sort((x, y) => x > y)
return [newArray[1], newArray[2]] // Or you could use your destructuring approach too!
```

In fact, if you look at the repo now, you'll see that I have updated it to do that: https://github.com/CodeYourFuture/js-exercises-tdd/blob/master/I.write-tests/05-get-second-third-smallest/get-second-third.js

File: I.write-tests/05-get-second-third-smallest/get-second-third.test.js (hunk 1 above)
Comment (rajgthub) [reply I.write-tests/05-get-second-third-smallest/get-second-third.test.js]: Fixed as above!

Comment (40thieves) [reply I.write-tests/05-get-second-third-smallest/get-second-third.test.js]: Just to be clear - your test here could be wrong still. For example, if the implementation code re-assigned the array that it was given to be an array of the same length but filled it with random numbers, then this test would still pass. The length of an array is not enough to check that it's contents are what you expect.

File: II.pass-tests/03-sum-numbers/add-numbers.js (hunk 2 above)
Comment (rajgthub) [reply II.pass-tests/03-sum-numbers/add-numbers.js]: module.exports = (numbers = [0]) => 
    numbers.length == 0 ? 0 : numbers.reduce((a, b) => a + b);
    expect(addAllnumbers()).toBe(0);//default value
    expect(addAllnumbers([])).toBe(0);

File: II.pass-tests/07-second-largest/second-largest.test.js (hunk 3 above)
Comment (rajgthub) [reply II.pass-tests/07-second-largest/second-largest.test.js]: No worries!

File: II.pass-tests/08-get-average/get-average.js (hunk 4 above)
Comment (rajgthub) [reply II.pass-tests/08-get-average/get-average.js]: ok.

File: II.pass-tests/08-get-average/get-average.js (hunk 5 above)
Comment (rajgthub) [reply II.pass-tests/08-get-average/get-average.js]: thanks

File: II.pass-tests/03-sum-numbers/add-numbers.js (hunk 2 above)
Comment (40thieves) [reply II.pass-tests/03-sum-numbers/add-numbers.js]: Firstly a quick github tip: if you put 3 backticks (```) on the line above a "block" of code then another 3 backticks on the line below. Then it will format the code nicely - it makes it easier to read.

Yep that would work! Nice work :)

There is another way that I was thinking of, when using the `.reduce` method. If you pass a second argument, it will use that to "initialise" the *accumulator* (which is the `a` variable in this example). So in this example the first time the callback function is called, `a` will be 0 and `b` will be 9. Then the next iteration, `a` will be 9 and `b` will be 23, etc. This is very useful in some cases.

File: I.write-tests/05-get-second-third-smallest/get-second-third.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (rajgthub) [reply I.write-tests/05-get-second-third-smallest/get-second-third.js]: That is interesting!

File: I.write-tests/05-get-second-third-smallest/get-second-third.test.js (hunk 1 above)
Comment (rajgthub) [reply I.write-tests/05-get-second-third-smallest/get-second-third.test.js]: ok

File: II.pass-tests/03-sum-numbers/add-numbers.js (hunk 2 above)
Comment (rajgthub) [reply II.pass-tests/03-sum-numbers/add-numbers.js]: thanks

Comment (rajgthub) [reply II.pass-tests/03-sum-numbers/add-numbers.js]: Sorry about that! I know it is hard to read. I was trying to figure out how to attach codes. Thanks for letting me know!

File: II.pass-tests/08-get-average/get-average.js (hunk 6 above)
Comment (rajgthub) [reply II.pass-tests/08-get-average/get-average.js]:
```exports.average = (elements = []) => {
    let len = 1;
    if (elements.length == 0) {
        return 0;
    }
    return elements.slice()
    .filter(element => typeof element === "number")
    .reduce((a, b, cindex) => {
        len = cindex + 1;
        return a + b}, 0) / len;
}```
```test('Average', function () {
    var numbers = [4, '-', 8, 11, 'hello', '57', 0, 2];
    var expected = 5;

    var output = average(numbers);
    expect(average()).toEqual(0);//default
    expect(average([])).toEqual(0);//empty
    expect(output).toEqual(expected);
});```

Comment (rajgthub) [reply II.pass-tests/08-get-average/get-average.js]: I think passing a test from you is hard :simple_smile:!

File: II.pass-tests/08-get-average/get-average.test.js (hunk 7 above)
Comment (rajgthub) [reply II.pass-tests/08-get-average/get-average.test.js]: ok

File: II.pass-tests/ES-6-practice/es6.js (hunk 8 above)
Comment (rajgthub) [reply II.pass-tests/ES-6-practice/es6.js]: Yes, we did not cover it. mmm....

Comment (40thieves) [reply II.pass-tests/ES-6-practice/es6.js]: In that case, don't worry about the other problems in this file. We'll cover them tomorrow or the week after

File: II.pass-tests/ES-6-practice/es6.js (hunk 9 above)
Comment (rajgthub) [reply II.pass-tests/ES-6-practice/es6.js]: Sorry! I have not read it properly! ```module.exports = {
  arrow: () => "es6 is awesome!",
  defaultParameter: (name = 'sam') => name,
  combineArrays: (arr1, arr2) => ([...arr1, ...arr2]),
  destructuring : ({cyf} = {}) => cyf,
  templateString: (a, b) => `The sum is equal to ${a + b}`
};```

File: II.pass-tests/ES-6-practice/es6.js (hunk 10 above)
Comment (rajgthub) [reply II.pass-tests/ES-6-practice/es6.js]: I have changed!

File: II.pass-tests/ES-6-practice/es6.js (hunk 11 above)
Comment (rajgthub) [reply II.pass-tests/ES-6-practice/es6.js]: Changed!

**Comment by rajgthub** [comment]
Thanks very much for your hard work and the micro-ish checking of the code...

File: III.tdd-katas/calculator-tdd/add.test.js [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    });
+});
+describe("Step 3: handle new lines between numbers", () => {
+    test("should return 26 for the string passed to add", () => {
+        expect(add("1,2,4fi5faswwe6,lgsf8")).toBe(26);
+    });
+});
+describe("Step 4: support different delimiters", () => {
+    test("should return 9.3 for the string passed to add", () => {
+        expect(add("//3[$5.3delimiter]\n[numbers..&.*1]")).toBeCloseTo(9.3);
```
Comment (textbook) [inline III.tdd-katas/calculator-tdd/add.test.js]: I don't think this test is correct - if the first line contains `//3[$5.3delimiter]` then the calculator should split the numbers in the second line based on *that whole string*, `3[$5.3delimiter]`. Look again at the example given: 

> For example `"//;\n1;2"` should return 3 where the default delimiter is `';'`.

File: III.tdd-katas/calculator-tdd/add.test.js [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    });
+    test('should return 1 for "1" string passed to add', () => {
+        expect(add("1")).toBe(1);
+    });
+    test('should return 3 for "1,2" string passed to add', () => {
+        expect(add("1,2")).toBe(3);
+    });
+});
+describe("Step 2: handle an unknown amount of numbers", () => {
+    test("should return 0 for empty string passed to add", () => {
```
Comment (textbook) [inline III.tdd-katas/calculator-tdd/add.test.js]: The test title here does not match the content of the test - I think this is a copy-paste error from above.

File: III.tdd-katas/calculator-tdd/add.test.js [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+        expect(add("1,2,4fi5faswwe6,lgsf8")).toBe(26);
```
Comment (textbook) [inline III.tdd-katas/calculator-tdd/add.test.js]: I wouldn't expect that; if the user's input contains things that *aren't* numbers separated by the delimiter, they should either be ignored or an error thrown (I'd prefer the latter). And if you are going to sanitise the input I'd expect the result to be 467, because if you remove the characters that aren't numbers or the delimiter you get `'1,2456,8'`.

File: III.tdd-katas/calculator-tdd/add.test.js [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    });
+});
+describe("Step 3: handle new lines between numbers", () => {
+    test("should return 26 for the string passed to add", () => {
+        expect(add("1,2,4fi5faswwe6,lgsf8")).toBe(26);
```
Comment (textbook) [inline III.tdd-katas/calculator-tdd/add.test.js]: This step describes handing newlines, but your test input doesn't contain any.

Comment (rajgthub) [reply III.tdd-katas/calculator-tdd/add.test.js]: I am extremely sorry for not replying to your comments. Thanks very much for such a long comments and I will do the changes as you pointed out. It was a fantastic class last week and learnt a lot of new things, especially the approach of TDD you showed step by step with the real example!
